equinox/_module/_module.py

```diff
@@ -0,0 +1,617 @@
+                        if converter_annotation is inspect.Signature.empty:
+                            converter_annotation = Any
+                cls.__init__.__annotations__[f.name] = converter_annotation
+        if has_dataclass_init:
+            cls.__init__.__doc__ = init_doc  # pyright: ignore[reportPossiblyUnboundVariable]
+
+        flattener = _ModuleFlattener(dataclasses.fields(cls))  # pyright: ignore[reportArgumentType]
+        jtu.register_pytree_with_keys(
+            cls,
+            flatten_with_keys=flattener.flatten_with_keys,  # pyright: ignore
+            flatten_func=flattener.flatten,  # pyright: ignore
+            unflatten_func=ft.partial(flattener.unflatten_with_cls, cls),  # pyright: ignore
+        )
+
+        return cls
+
+    def __call__(cls, *args, **kwargs):  # noqa: N805
+        __tracebackhide__ = True
+        if cls in _abstract_module_registry:
+            # Any other is-abstract checks will be handled in super().__call__.
+            raise TypeError("Cannot instantiate abstract `equinox.Module`.")
+        if _has_dataclass_init[cls]:
+            for x in jtu.tree_leaves((args, kwargs)):
+                _warn_jax_transformed_function(cls, x)
+
+        self = None
+        try:
+            self = super().__call__(*args, **kwargs)  # pyright: ignore[reportAttributeAccessIssue]
+        finally:
+            if self is not None:
+                _currently_initialising.remove(self)
+        assert not is_abstract_module(cls)  # pyright: ignore[reportArgumentType]
+
+        fields = dataclasses.fields(cls)  # pyright: ignore[reportArgumentType]
+        missing_names = {
+            f.name
+            for f in fields
+            # Not `vars` or `__dict__`, to allow for `property`s overwriting a field.
+            # Not recommended, but allowable for backward compatibility.
+            if f.name not in dir(self)
+        }
+        if len(missing_names) > 0:
+            raise TypeError(
+                f"The following fields were not initialised during __init__: "
+                f"{missing_names}"
+            )
+
+        for f in fields:
+            if (converter := f.metadata.get("converter")) is not None:
+                object.__setattr__(self, f.name, converter(getattr(self, f.name)))
+            if f.metadata.get("static", False):
+                if any(jtu.tree_map(is_array, jtu.tree_leaves(getattr(self, f.name)))):
+                    warnings.warn(
+                        "A JAX array is being set as static! This can result "
+                        "in unexpected behavior and is usually a mistake to do.",
+                        stacklevel=2,
+                    )
+
+        for parent_cls in cls.__mro__:
+            try:
+                check_init = parent_cls.__dict__["__check_init__"]
+            except KeyError:  # noqa: PERF203
+                pass
+            else:
+                check_init(self)
```

Comment:
```suggestion
        for parent_cls in cls.__mro__:
            if (check_init := parent_cls.__dict__.get("__check_init__", False)):
                check_init(self)
```

Reply:
Hmm, this is relying on an implicit-cast-to-bool (in this case that functions are truthy). I try to avoid ever using this, I regard it as a misfeature / an easy source of bugs.

Reply:
What about an is not False check than?

Reply:
... Or `parent_cls.__dict__.get("__check_init__")) is not None:`